circuit_passes/src/bucket_interpreter/env/function_env.rs

```diff
@@ -0,0 +1,235 @@
+        &self,
+        new_frame: CallStackFrame,
+    ) -> Option<CallStack> {
+        if !self.call_stack.contains(&new_frame) && self.call_stack.depth() <= CALL_STACK_LIMIT {
+            Some(self.call_stack.clone().push(new_frame))
+        } else {
+            None
+        }
+    }
+
+    pub fn get_caller_stack(&self) -> &[BucketId] {
+        &self.caller_stack
+    }
+
+    pub fn get_var(&self, idx: usize) -> Value {
+        self.vars.get(&idx).unwrap_or_default().clone()
+    }
+
+    pub fn get_signal(&self, _idx: usize) -> Value {
+        // There are no signals in source functions.
+        unreachable!()
+    }
+
+    pub fn get_subcmp_signal(&self, _subcmp_idx: usize, _signal_idx: usize) -> Value {
+        // There are no components in source functions.
+        unreachable!()
+    }
+
+    pub fn get_subcmp_name(&self, _subcmp_idx: usize) -> &String {
+        // There are no components in source functions.
+        unreachable!()
+    }
+
+    pub fn get_subcmp_template_id(&self, _subcmp_idx: usize) -> usize {
+        // There are no components in source functions.
+        unreachable!()
+    }
+
+    pub fn get_subcmp_counter(&self, _subcmp_idx: usize) -> Value {
+        // There are no components in source functions.
+        unreachable!()
+    }
+
+    pub fn subcmp_counter_is_zero(&self, _subcmp_idx: usize) -> bool {
+        // There are no components in source functions.
+        unreachable!()
+    }
+
+    pub fn subcmp_counter_equal_to(&self, _subcmp_idx: usize, _value: usize) -> bool {
+        // There are no components in source functions.
+        unreachable!()
+    }
+
+    pub fn get_vars_sort(&self) -> BTreeMap<usize, Value> {
+        sort(&self.vars, Clone::clone)
+    }
+
+    pub fn collect_write(
+        &self,
+        dest_address_type: &AddressType,
+        idx: usize,
+        collector: &mut Writes,
+    ) {
+        match dest_address_type {
+            AddressType::Variable => collector.vars.as_mut().map(|s| s.insert(idx)),
+            AddressType::Signal => collector.signals.as_mut().map(|s| s.insert(idx)),
+            AddressType::SubcmpSignal { .. } => unreachable!("Source function cannot have subcmp"),
+        };
+    }
+
+    // WRITE OPERATIONS
+    pub fn set_var(self, idx: usize, value: Value) -> Self {
+        let mut copy = self;
+        copy.vars.insert(idx, value);
+        copy
+    }
+
+    pub fn set_signal(self, _idx: usize, _value: Value) -> Self {
+        // There are no signals in source functions.
+        unreachable!()
+    }
+
+    pub fn set_vars_to_unknown<T: IntoIterator<Item = usize>>(self, idxs: Option<T>) -> Self {
+        let mut copy = self;
+        if let Some(idxs) = idxs {
+            for idx in idxs {
+                copy.vars.insert(idx, Value::Unknown);
+            }
+        } else {
+            for (_, v) in copy.vars.iter_mut() {
+                *v = Value::Unknown;
+            }
+        }
```

Comment:
Why is the `None` behavior to write all the variables to Unknown? Is it for soundness?

Reply:
Yes. If the interpreter cannot compute the BranchBucket condition, then it does not know which block of the BranchBucket will execute. Thus, all memory locations that COULD be written by EITHER block must be set to Unknown.